### PR TITLE
Refactor `MLP` class and delegate encoding to `Predictor` 

### PR DESCRIPTION
**Pull request by shihchengli**
## Description
As described in #754, the `MPNN.encoding()` is broken. This PR introduces two modifications to resolve the issue:

(1) It restructures the `MLP` from a sequence of "layers" to a sequence of "blocks". Additionally, it introduces a class method, `MLP.build()`, which should be invoked to create a new `MLP` object. This alteration allows the blocks to be sliced using specified indices, as an `nn.Sequential` object.
(2) It transfers the encoding responsibilities to the `Predictor`. This change assigns the task of calculating an "encoding" to the individual predictor. The encoding is defined as the "final hidden representation of the FFN", implying that it is inherently a function determined by the predictor rather than being an attribute at the `MPNN` level.

## Example / Current workflow
```
import pandas as pd
import numpy as np
import torch
from lightning import pytorch as pl
from chemprop import data, featurizers, models

# Load model
checkpoint_path = '../tests/data/example_model_v2_regression_mol.ckpt' # path to the checkpoint file. 
mpnn = models.MPNN.load_from_checkpoint(checkpoint_path)

# Load data
test_path = '../tests/data/smis.csv'
smiles_column = 'smiles'
df_test = pd.read_csv(test_path)
smis = df_test[smiles_column]
test_data = [data.MoleculeDatapoint.from_smi(smi) for smi in smis]
featurizer = featurizers.SimpleMoleculeMolGraphFeaturizer()
test_dset = data.MoleculeDataset(test_data, featurizer=featurizer)
test_loader = data.MolGraphDataLoader(test_dset, shuffle=False)

# Get encoding
all_fingerprints = []
with torch.no_grad():
    for batch in test_loader: 
        bmg, V_d, X_d, *_ = batch  # Update these based on your batch structure
        fingerprints = mpnn.encoding(bmg, V_d, X_d)
all_fingerprints.append(fingerprints)
```
```
>> [tensor([[ 0.0466, -0.6394,  0.7126,  ..., -0.1838, -0.4809, -0.5542],
         [-0.1257,  0.3576, -0.7796,  ...,  0.0377, -0.4563,  0.0618],
         [-0.2197, -0.0084,  0.0840,  ...,  0.2293, -0.0562, -0.2252],
         ...,
         [ 0.0515,  0.6676, -0.4578,  ...,  0.0615, -0.4609,  0.2687],
         [ 0.1989,  0.0521,  0.5110,  ...,  0.1003, -1.3129,  0.0610],
         [ 0.1516, -0.0287, -0.1152,  ..., -0.2354,  0.0292, -0.0623]])]
```

## Questions
In v1, I believe the output for "the final hidden" is derived from the input of the final layer. That is, the definition also applies an activation function and a dropout before returning them. This PR modifies the behavior to return the raw linear outputs instead.
```
# Example
Sequential(
  (0): Linear(in_features=64, out_features=128, bias=True)
  (1): ReLU()
  (2): Dropout(p=0.1, inplace=False)
  (3): Linear(in_features=128, out_features=128, bias=True)  # this PR intend to use this output as "the final hidden"
  (4): ReLU()
  (5): Dropout(p=0.1, inplace=False)  # v1 uses this output as "the final hidden"
  (6): Linear(in_features=128, out_features=2, bias=True)
)
```
This change could enhance expressivity, as the output scale is no longer constrained by the activation function. If devs have any thoughts on this change, please feel free to provide some input.

## Relevant issues
#754 

## Checklist
- [ ] linted with flake8?
- [ ] (if appropriate) unit tests added?

### CODE REVIEW

**Comment by davidegraff** [comment]
N.b., to keep track of the thread, these changes were first detailed [here](https://github.com/chemprop/chemprop/issues/754#issuecomment-2034814931).

**Review by davidegraff — APPROVED** [review]
lgtm

**Comment by davidegraff** [comment]
@shihchengli please do not merge without two approving reviews.

**Comment by shihchengli** [comment]
I am sorry. I will reopen the PR after #759 is merged in.
